Improve docs (#20)
cqse/teamscale-cli: README.md

The [Teamscale](https://teamscale.com) pre-commit command line interface allows you to integrate pre-commit analysis in editors or IDEs such as VS Code, Emacs, QtCreator or Sublime by providing findings in a standard error format that is interpreted and displayed like compile errors by these editors.

## Installation

**This guide uses `pip` and `python`. If you prefer Python 3, you'll probably have to replace all occurrences with `pip3` and `python3`.**

1. Install `libgit2` (https://libgit2.org) on all platforms except Windows
    - On Ubuntu use `apt-get install libgit2-dev`.
    - On macOS use `brew install libgit2`.
    - Not necessary on Windows as `libgit2` is already packaged with the `pygit` dependency.

2. Install `teamscale-cli` via pip:

    ```bash
    $ pip install teamscale-cli
    ```

>     The `teamscale-cli` executable will be installed to `~/.local/bin/teamscale-cli`. Please ensure it is on your `PATH`.

Reviewer: I think something went wrong here. I'd keep 4. and move 5. below the bash code snippet.

# Teamscale Pre-Commit Command Line Client [![Build Status](https://travis-ci.org/cqse/teamscale-cli.svg?branch=master)](https://travis-ci.org/cqse/teamscale-cli) [![PyPI version](https://badge.fury.io/py/teamscale-cli.svg)](https://badge.fury.io/py/teamscale-cli) [![Teamscale Project](https://img.shields.io/badge/teamscale-teamscale--cli-brightgreen.svg)](https://demo.teamscale.com/activity.html#/teamscale-cli)

The [Teamscale](https://teamscale.com) pre-commit command line interface allows you to integrate pre-commit analysis in editors or IDEs such as VS Code, Emacs, QtCreator or Sublime by providing findings in a standard error format that is interpreted and displayed like compile errors by these editors.

## Installation

> **This guide uses `pip` and `python`. If you prefer Python 3, you'll probably have to replace all occurrences with `pip3` and `python3`.**

Reviewer: Is the default for python/pip python2 on most distros python2? Asking because on Arch Linux python3 is the default.

Author: My Ubuntu still has python2. I'm guessing RedHat won't be any more up-to-date ;-)

1. Install `libgit2` (https://libgit2.org) on all platforms except Windows
    - On Ubuntu use `apt-get install libgit2-dev`.
    - On macOS use `brew install libgit2`.
    - Not necessary on Windows as `libgit2` is already packaged with the `pygit` dependency.

2. Install `teamscale-cli` via pip:

```bash
$ pip install teamscale-cli

> ```

Reviewer: ```suggestion
	```bash
	$ pip install teamscale-cli
	```
```

Reviewer: Indentation, so that it's visually under point 2.

> The `teamscale-cli` executable will be installed to `~/.local/bin/teamscale-cli`. Please ensure it is on your `PATH`.

Reviewer: ```suggestion
	The `teamscale-cli` executable will be installed to `~/.local/bin/teamscale-cli`. Please ensure it is on your `PATH`.
```

Reviewer: Indentation, so that it's visually under point 2.

3. Copy the example configuration file [.teamscale-precommit.config](./config/.teamscale-precommit.config) to the root directory of the repository you want to analyze. Edit it and insert all the necessary data. You can find your personal access token by opening Teamscale and clicking on your Avatar in the top right corner.

> 4. Use `teamscale-cli` as a compile or build command in your editor. See below for instructions for popular editors. Provide a file or folder within your repository as input. The general invocation looks like this:

Reviewer: ```suggestion
4. Use `teamscale-cli` as a compile or build command in your editor. See [below](#instructions-for-popular-editors) for instructions for popular editors. Provide a file or folder within your repository as input. The general invocation looks like this:
```

```bash
$ teamscale-cli [OPTIONS] CURRENTLY_OPENED_EDITOR_FILE

> ```

Reviewer: ```suggestion
	```bash
	$ teamscale-cli [OPTIONS] CURRENTLY_OPENED_EDITOR_FILE
	```
```

Reviewer: Indentation, so that it's visually under point 4.

## How it works

> teamscale-cli will use the path you specify on the command line to locate your .git directory. It then uploads all [local uncommitted changes known to Git](#how-does-change-detection-work) from your Git repository to the Teamscale server and project you configured in the config file. The client then waits until Teamscale has analyzed these changes, and outputs findings on `stdout` in the following format:

Reviewer: ```suggestion
`teamscale-cli` will use the path you specify on the command line to locate your .git directory. It then uploads all [local uncommitted changes known to Git](#how-does-change-detection-work) to the Teamscale server and project you configured in the config file. The client then waits until Teamscale has analyzed these changes, and outputs findings on `stdout` in the following format:
```
I think it's clear that we only upload changes from the current repository if we say "changes known to Git".

Also, back ticks around `teamscale-cli`

```
Path to file:line number:Column: (warning|error): Message
```

> This allows you to use the highlighting capabilities of your editors to highlight lines with findings in the editor and to jump to the finding location from the pre-commit console output. For your convenience, we've provided sample configurations for some editors and IDEs below.

Reviewer: ```suggestion
This allows you to use the highlighting capabilities of your editors to highlight lines with findings in the editor and to jump to the finding location from the pre-commit console output. For your convenience, we've provided sample configurations for some editors and IDEs [below](#instructions-for-popular-editors).
```

Which findings are output is controlled by several command line options:

```
teamscale-cli CURRENTLY_OPENED_EDITOR_FILE
```

will output only pre-commit findings, i.e. findings that were added or removed due to your local changes. This includes all changes known to Git, not just the `CURRENTLY_OPENED_EDITOR_FILE`. Existing findings will not be output.
Use this if you are only interested in the impact your local code changes will have on the findings.

```
teamscale-cli --fetch-existing-findings CURRENTLY_OPENED_EDITOR_FILE
```

will output pre-commit findings for all locally changed files plus all existing, unchanged findings in `CURRENTLY_OPENED_EDITOR_FILE`.
Use this if you also want to have a look at existing findings in the currently opened file, e.g. to clean up old findings while you code.

> **We recommend you use this mode as you can use it to see both the existing findings in the currently opened file (regardless of whether it is changed locally or not and the impact of your local changes before you commit them. This gives you the opportunity to look at existing findings in the files you open and clean some of them up while you work on your code.**

Reviewer: ```suggestion
**We recommend you use this mode as you can use it to see both the existing findings in the currently opened file (regardless of whether it is changed locally or not) and the impact of your local changes before you commit them. This gives you the opportunity to look at existing findings in the files you open and clean some of them up while you work on your code.**
```
Missing closing bracket.

```
teamscale-cli --fetch-existing-findings-in-changes CURRENTLY_OPENED_EDITOR_FILE
```

will output pre-commit findings for all locally changed files plus all existing, unchanged findings in all locally changed files.
Use this if you always want to clean up in all files where you have made changes.

Run the client with the `-h` argument to see additional available options.

> ## Troubleshooting

Reviewer: I'd move Troubleshooting to the end of the file.

## Troubleshooting

- If python does not find the name `ConverterMapping` try uninstalling the `python-configparser` system package and install `configparser` via pip.

## Instructions for Popular Editors

### Sublime

Add a new *Build System* under `Tools > Build System`. Locate `config/teamscale-precommit.sublime-build` in this repo. Copy and paste the snippet and modify the arguments to fit your needs.
Add a new *Build System* under `Tools > Build System`. Copy and paste [our example snippet](./config/teamscale-precommit.sublime-build) and modify the arguments to fit your needs.

### Xcode

Add a new *Build Phase* (`New Run Script Phase`) to your project. Enter the following command as shell script in that phase (see screenshot):

```bash
teamscale-cli ${SRCROOT} --fail-on-red-findings
python -c 'from teamscale_precommit_client.precommit_client import run;run()' ${SRCROOT}
```

**In Xcode you cannot use `--fetch-existing-findings` since there is unfortunately no environment variable for the currently opened editor file.
Using this flag would give you all findings in the entire project, which is usually not helpful.**

![Configuring the Build Phase in Xcode](config/xcode_1.png)

> As you're on a Mac, make sure to use the correct Python version in that snippet, which might be `python3`. The option `--fail-on-red-findings` will fail your Xcode build if new RED findings were found. You might decide to drop that flag. If you've done it right, Xcode will show all findings inlined as seen on the following screenshot:

Reviewer: If we remove this, I think we also need to adapt the screenshot to avoid confusion (it's showing the `fail-on-red-findings` flag).

Author: I'll ping Dennis to do this, since he's set this up already :-) but I'll leave it for another PR because I don't know when he'll find the time

> This should allow you to run `:compiler teamscale` and `:make %`. Then you should be able to use your usual workflow (e.g. `:cn`) to go through the findings.

Reviewer: ```suggestion
This should allow you to run `:compiler teamscale` and `:make %`. Then, you should be able to use your usual workflow (e.g. `:cn`) to go through the findings.
```
Missing comma.

> ### QtCreator

Reviewer: ```suggestion
### Qt Creator
```
There is a space in between Qt and Creator according to [the Qt website](https://www.qt.io/product/development-tools). Misspelled it many times myself :D

> In order to use the pre-commit Analysis in QtCreator you need to add a new Kit (Teamscale does _not_ have to be the default Kit).

Reviewer: ```suggestion
In order to use the pre-commit Analysis in Qt Creator you need to add a new Kit (Teamscale does _not_ have to be the default Kit).
```

Add a new build configuration:

> ![Pre-Commit](config/qtcreator_2.png)

Reviewer: Maybe a better option for the screenshot would be [here](https://github.com/cqse/teamscale-cli/blob/update_qt_creator_docu/config/qtcreator_2.png). I find it weird to have one screenshot with light and one with dark theme.

Reviewer: Also, I think we can discard #14 then, right?

Author: whoops, I totally forgot that that was still open. Also didn't see your comment there about merging... I'll go over your changes there and make sure they're all incorporated here. My bad :-(

Author: Acutally, I prefer your dark screenshot because it also highlights how to run the multiple configs. Could you instead create a dark theme screenshot of the compiler settings? then I'll replace the first screenshot to have both in dark theme :-) @AnonymFx

Author: merged #14 and then merged master into this branch :-)

Reviewer: I actually preferred replacing the dark theme screenshot as it was not optimal anyways in my opinion. Please check if that's fine for you as well :)

Author: It still shows the dark screenshot for me? https://github.com/cqse/teamscale-cli/blob/improve_docs/config/qtcreator_2.png

Reviewer: Nope, that's actually the new one. Compare to https://github.com/cqse/teamscale-cli/blob/6eb45ba9113c66c94a56eaa699390a610a8251d0/config/qtcreator_2.png

Reviewer: I guess the remaining dark elements are dark even in a light qt theme

Author: I see, weird :P works for me though

Then configure it as follows:  

* Executable: `python`
* Command line arguments: `-c "from teamscale_precommit_client.precommit_client import run;run()" --fetch-existing-findings --log-to-stderr %{CurrentDocument:FilePath}`

The flag `--log-to-stderr` is required as otherwise QtCreator will not recognize the findings.  

> You can add more than one build configuration if you want to e.g. sometimes run the teamscale-cli with different arguments.

Reviewer: ```suggestion
You can add more than one build configuration, e.g. if you want to sometimes run the teamscale-cli with different arguments.
```

2. Install `teamscale-cli` via pip:

    ```bash
    $ pip install teamscale-cli
    ```

    The `teamscale-cli` executable will be installed to `~/.local/bin/teamscale-cli`. Please ensure it is on your `PATH`.

3. Copy the example configuration file [.teamscale-precommit.config](./config/.teamscale-precommit.config) to the root directory of the repository you want to analyze. Edit it and insert all the necessary data. You can find your personal access token by opening Teamscale and clicking on your Avatar in the top right corner.

> 5. The behavior of the client can be tweaked with several arguments. Run the client with the `-h` argument to get the usage.

Reviewer: I think something went wrong there. It'd make more sense if we keep 4. and move 5. below the bash code snippet.

Author: -h is already mentioned it the end of the next section. I restored 4. as suggested :-)

New files that are not in the index will be ignored.

> ## Other useful command line options

Reviewer: Why remove some of the command line options here but not all of them?
I'd either keep or remove both. My preference would be to remove both in order since the exact information can be retrieved by running `teamscale-cli -h`.

Author: good point
